metrics/README.md

```diff
@@ -0,0 +1,99 @@
+# Common metrics export interfaces for Knative
+
+_Note that this directory is currently in transition. See [the Plan](#the-plan)
+for details on where this is heading._
+
+## Current status
+
+The code currently uses OpenCensus to support exporting metrics to multiple
+backends. Currently, two backends are supported: Prometheus and Stackdriver.
+
+Metrics export is controlled by a ConfigMap called `config-observability` which
+is a key-value map with specific values supported for each of the Stackdriver
+and Prometheus backends. Hot-reload of the ConfigMap on a running process is
+supported by directly watching (via the Kubernetes API) the
+`config-observability` object. Configuration via environment is also supported
+for use by the `queue-proxy`, which runs with user permissions in the user's
+namespace.
+
+## Problems
+
+There are currently
+[6 supported Golang exporters for OpenCensus](https://opencensus.io/exporters/supported-exporters/go/).
+At least the Stackdriver exporter causes problems/failures if started without
+access to (Google) application default credentials. It's not clear that we want
+to build all of those backends into the core of `knative.dev/pkg` and all
+downstream dependents, and we'd like all the code shipped in `knative.dev/pkg`
+to be able to be tested without needing any special environment setup.
+
+With the current direct-integration setup, there needs to be initial and ongoing
+work in `pkg` (which should be high-value, low-churn code) to maintain and
+update stats exporters which need to be statically linked into ~all Knative
+binaries. This setup also causes problems for vendors who may want or need to
+perform an out-of-tree integration (e.g. proprietary or partially-proprietary
+monitoring stacks).
+
+Another problem is that each vendor's exporter requires different parameters,
+supplied as Golang `Options` methods which may require complex connections with
+the Knative ConfigMap. Two examples of this are secrets like API keys and the
+Prometheus monitoring port (which requires additional service/etc wiring).
+
+See also
+[this doc](https://docs.google.com/document/d/1t-aov3XrhobjCKW4kwScY44QAoahiwxoyXXFtZyL8jw/edit),
+where the plan was worked out.
+
+## The plan
+
+OpenCensus (and eventually OpenTelemetry) offers an sidecar or host-level agent
+with speaks the OpenCensus protocol and can proxy from this protocol to multiple
+backends.
+
+![OpenCensus Agent configuration](https://github.com/census-instrumentation/opencensus-service/raw/master/images/opencensus-service-deployment-models.png)
+(From OpenCensus Documentation)
+
+**We will standardize on export to the OpenCensus export protocol, and encourage
+vendors to implement their own OpenCensus Agent or Collector DaemonSet, Sidecar,
+or other `localhost:55678` service which connects to their desired monitoring
```

Comment:
Hmm, I'm not sure I understand why Knative would want to support the OpenCensus Agent. It seems like Knative should force metrics to go through the centralized collector service, and do a bring-your-own-collector-service model. Otherwise, we'll have the issues of short-lived queue-proxy pods dropping metrics if we allow OpenCensus Agent sidecars to export metrics directly.

Wouldn't the ideal state be for Knative controllers to be like the task + Open Census library (embedded) component of this diagram? https://github.com/census-instrumentation/opencensus-service/blob/master/DESIGN.md#architecture-overview-1

Reply:
I don't want to specify what the ideal state might be, particularly for the case of things like Cloud Run (fully managed) that might end up running their own implementation of an OpenCensus receiver.

Assuming that you're using Prometheus or Stackdriver, I would imagine implementing this as a node-level service (e.g. a DaemonSet, rather than a sidecar) such that:

1. All the containers can simply talk to `localhost` and don't need to agree on a name. (Also, they can just use the default from the OpenCensus library.)
2. There's only a single network hop for collection -- for Prometheus, this would be a fixed list of all the nodes in the system; for Stackdriver, this would be localhost -> agent -> off-cluster.
3. It would be fine to run the collector instead of the agent if you needed something that the collector did. It seemed like the collector might be a little more complicated, but some operators might need that power.

Reply:
Hmm, on further review I remain puzzled by the functional differences between the OpenCensus Agent and the OpenCensus Collector service if the Agent can still receive from multiple sources. I agree with you that Knative does not need to specify Agent vs. Collector for vendors.

What I really meant was should Knative allow sidecar as an option? It would be pretty difficult to vendors to bring their own agents as sidecars if Knative does not provide an easy path for queue-proxy sidecars. It also brings back the short-lived pod issue and concerns about sidecar perf like with the FluentD one.